src/persist/src/indexed/mod.rs

```diff
@@ -285,18 +234,31 @@ impl AppliedState {
             graveyard: meta.graveyard,
             unsealeds,
             traces,
-            prev_meta: meta_copy,
         }
     }
+
+    fn assign_seqno(&mut self) -> SeqNo {
+        let seqno = self.highest_assigned_seqno + 1;
+        self.highest_assigned_seqno = seqno;
+        seqno
+    }
 }
 
 impl<L: Log, B: Blob> Indexed<L, B> {
     /// Attempt to commit the current in-memory metadata state to durable storage,
     /// and if not, revert back to a previous version.
-    fn try_set_meta(&mut self, prev_meta: BlobMeta) -> Result<(), Error> {
-        debug_assert_eq!(self.validate_matches_storage(&prev_meta), Ok(()));
+    ///
+    /// Precondition: pending has been emptied
+    fn try_set_meta(&mut self, prev_meta: BlobMeta, new_meta: BlobMeta) -> Result<(), Error> {
```

Comment:
maybe I'm missing something but why did we change `try_set_meta` to take both prev and new as arguments? Seems like every callsite now has to call `self.state.serialize_meta()` before calling this (at least, from the callsites directly visible on the github ui)

Reply:
Hmm, I think the real answer is that for most of the life of this refactor, I had moved highest_assigned_seqno to Pending, which meant that the different callers had to deal with whether Pending was Some or None when calling serialize_meta, but then I abandoned that at the last second because I realized that the logic is much cleaner without it moved to Pending. The more important thing is where it most correctly "belongs" but tbh I haven't quite convinced myself of anything there yet.

Reply:
all I'm trying to say is that for me, this diff

```
diff --git a/src/persist/src/indexed/mod.rs b/src/persist/src/indexed/mod.rs
index b462201a8..e732cd54e 100644
--- a/src/persist/src/indexed/mod.rs
+++ b/src/persist/src/indexed/mod.rs
@@ -248,7 +248,7 @@ impl<L: Log, B: Blob> Indexed<L, B> {
     /// and if not, revert back to a previous version.
     ///
     /// Precondition: pending has been emptied
-    fn try_set_meta(&mut self, prev_meta: BlobMeta, new_meta: BlobMeta) -> Result<(), Error> {
+    fn try_set_meta(&mut self, prev_meta: BlobMeta) -> Result<(), Error> {
         // NB: This validate_pending_empty is intentionally a returned error
         // instead of an assert because it's a precondition (and so a violation
         // means a usage error by the caller of this).
@@ -258,6 +258,7 @@ impl<L: Log, B: Blob> Indexed<L, B> {
             Ok(())
         );

+        let new_meta = self.state.serialize_meta();
         // TODO: Instead of fully overwriting META each time, this should be
         // more like a compactable log.
         if let Err(e) = self.blob.set_meta(&new_meta) {
@@ -321,9 +322,7 @@ impl<L: Log, B: Blob> Indexed<L, B> {
                 return Err(err);
             }
         };
-        let new_meta = self.state.serialize_meta();
-
-        self.try_set_meta(meta_before, new_meta)?;
+        self.try_set_meta(meta_before)?;

         debug_assert_eq!(self.validate_pending_empty(), Ok(()));
         debug_assert_eq!(self.validate(), Ok(()));
@@ -595,8 +594,7 @@ impl<L: Log, B: Blob> Indexed<L, B> {
                 self.state = AppliedState::new(meta_before);
                 Err(err)
             } else {
-                let new_meta = self.state.serialize_meta();
-                self.try_set_meta(meta_before, new_meta)
+                self.try_set_meta(meta_before)
             }
         })();
```

+ maybe some docs updates around `try_set_meta` to make it clear that it serializes out the current state and tries to save it to storage seems like a simpler, harder to misuse api?

Reply:
Done